src/project.rs

```diff
@@ -0,0 +1,325 @@
+    // the contents of the lock file
+    pub fn show_deps(&self) -> Result<(), Error> {
+        self.venv_runner.run("pip", vec!["list"])
+    }
+
+    /// Show the resolved virtualenv path.
+    //
+    // See `PathsResolver.paths()` for details
+    pub fn show_venv_path(&self) -> Result<(), Error> {
+        println!("{}", self.paths.venv.display());
+        Ok(())
+    }
+
+    /// Same has `show_venv_path`, but add the correct subfolder
+    /// (`bin` on Linux and macOS, `Scripts` on Windows).
+    pub fn show_venv_bin_path(&self) -> Result<(), Error> {
+        let bin_path = self.venv_runner.binaries_path();
+        println!("{}", bin_path.display());
+        Ok(())
+    }
+
+    pub fn show_outdated(&self) -> Result<(), Error> {
+        self.venv_runner
+            .run("pip", vec!["list", "--outdated", "--format", "columns"])
+    }
+
+    fn install_editable(&self) -> Result<(), Error> {
+        let mut message = "Installing deps from setup.py".to_string();
+        if self.settings.production {
+            message.push_str(" using 'prod' extra dependencies");
+        } else {
+            message.push_str(" using 'dev' extra dependencies");
+        }
+        print_info_2(&message);
+
+        let mut args = vec!["-m", "pip", "install", "--editable"];
+        if self.settings.production {
+            args.push(".[prod]")
+        } else {
+            args.push(".[dev]")
+        }
+        self.venv_runner.run("python", args)
+    }
+
+    // Actually write the lock file
+    // Delegates most of the work to the Lock struct.
+    fn write_lock(&self, lock_options: &operations::LockOptions) -> Result<(), Error> {
+        let metadata = &self.get_metadata()?;
+        let frozen_deps = self.get_frozen_deps()?;
+        let lock_path = &self.paths.lock;
+        operations::write_lock(lock_path, frozen_deps, lock_options, metadata)
+    }
+
+    fn get_metadata(&self) -> Result<Metadata, Error> {
+        let dmenv_version = env!("CARGO_PKG_VERSION");
+        let python_platform = &self.python_info.platform;
+        let python_version = &self.python_info.version;
+        Ok(Metadata {
+            dmenv_version: dmenv_version.to_string(),
+            python_platform: python_platform.to_string(),
+            python_version: python_version.to_string(),
+        })
+    }
+
+    /// Get the list of the *actual* deps in the virtualenv by calling `pip freeze`.
+    fn get_frozen_deps(&self) -> Result<Vec<FrozenDependency>, Error> {
+        let freeze_output = self.run_pip_freeze()?;
+        let mut res = vec![];
+        for line in freeze_output.lines() {
+            let frozen_dep = FrozenDependency::from_string(&line)?;
+            // Filter out pkg-resources. This works around
+            // a Debian bug in pip: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=871790
+            if frozen_dep.name != "pkg-resources" {
+                res.push(frozen_dep);
+            }
+        }
```

Comment:
The same in functional style, just because for loops are icky:
```rust
// Filter out pkg-resources. This works around
// a Debian bug in pip: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=871790
let filter = |dep: &FrozenDependency| dep.name != "pkg-resources";
freeze_output.lines()
    .map(FrozenDependency::from_string)
    .filter(|result| result.as_ref().map(filter).unwrap_or(true))
    .collect()
```